### PR TITLE
Pass in aws client

### PR DESCRIPTION
**Pull request by albertrdixon**
~AWS credentials can come from a bunch of sources. Rather than only supporting passing credentials from the caller, let's be a bit more flexible:~

- ~If given a role arn, then use AssumedRole creds, using passed credentials if supplied~
- ~If just given creds, use them directly~
- ~Fallback to using SharedCredentials and have the AWS client figure it out from there~

Allow the caller to pass in an s3 client so non-standard setups can be supported ad nauseam

### CODE REVIEW

File: lib/prebundler/s3_backend.rb [hunk 1]
```diff
@@ -33,15 +40,15 @@ def list_files
...
+        prefix: "#{Bundler.local_platform}/#{Prebundler.platform_version}/#{Gem.extension_api_version}"
       }
 
       while truncated
         options = if continuation_token
-          base_options.merge(continuation_token: continuation_token)
-        else
-          base_options
-        end
+                    base_options.merge(continuation_token: continuation_token)
```
Comment (camertron) [inline lib/prebundler/s3_backend.rb]: I know this is Lumos style, but I personally much prefer only a single level of indentation :)

**Review by camertron — CHANGES_REQUESTED** [review]
Hmm the number of options you can pass into `S3Backend` is getting pretty large. What would you think of passing the `Aws::S3::Client` into the constructor instead? I.e.

```ruby
s3 = Aws::S3::Client.new(...)

Prebundler.configure do |config|
  config.storage_backend = Prebundler::S3Backend.new(s3)
end
```

File: lib/prebundler/s3_backend.rb (hunk 1 above)
Comment (albertrdixon) [reply lib/prebundler/s3_backend.rb]: damn editors....

**Comment by albertrdixon** [comment]
you know, I kind of thought you were going to say that. It did look a little ugly there for gathering creds. Only downside is that one would need to know how to configure the ruby aws-sdk in order to use prebundler /shrug

**Comment by camertron** [comment]
Yeah I think that's fine. Just seems more straightforward to let the AWS gem do what it's good at.

**Comment by camertron** [comment]
@albertrdixon huh I see your commits but the "Files changed" view hasn't changed. WTF github.

**Comment by albertrdixon** [comment]
@camertron looks like its updated now (All this PR is really doing now is adding `client` as an option to pass to `.new`

File: lib/prebundler/s3_backend.rb [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
     def initialize(options = {})
-      @access_key_id = options.fetch(:access_key_id)
-      @secret_access_key = options.fetch(:secret_access_key)
-      @bucket = options.fetch(:bucket)
-      @region = options.fetch(:region) { 'us-east-1' }
-      @endpoint = options.fetch(:endpoint) { 's3.amazonaws.com' }
-      @force_path_style = options.fetch(:force_path_style) { false }
+      @bucket             = options.fetch(:bucket)
+
+      @client             = options.fetch(:client, nil)
```
Comment (camertron) [inline lib/prebundler/s3_backend.rb:12]: If we're passing the client in, do we need to continue passing the other options too (i.e. `access_key_id` and friends)?

**Review by camertron — APPROVED** [review]

File: lib/prebundler/s3_backend.rb (hunk 2 above)
Comment (camertron) [reply lib/prebundler/s3_backend.rb:12]: Never mind, this allows not only backwards-compatibility but also makes standard setups easy.

**Comment by camertron** [comment]
Published as v0.11.0.

**Pull request merged**